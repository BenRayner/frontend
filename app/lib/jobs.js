export function jobTime(job) {
  const { state, startedAt, finishedAt, canceledAt, timedOutAt } = job;
  const time = {};

  // job never stared, so no duration to show
  if (!startedAt) {
    return time;
  }

  time.from = startedAt;

  switch (state) {
    case 'finished':
      time.to = finishedAt;
      break;

    case 'canceled':
    case 'canceling':
      time.to = canceledAt;
      break;

    case 'timed_out':
    case 'timing_out':
      time.to = timedOutAt;
      break;
  }

  return time;
}
