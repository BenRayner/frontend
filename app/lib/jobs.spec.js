/* global describe, it, expect */
import { jobTime } from './jobs';

const JOB_STATES = [
  'pending',
  'waiting',
  'waiting_failed',
  'blocked',
  'blocked_failed',
  'unblocked',
  'unblocked_failed',
  'scheduled',
  'assigned',
  'accepted',
  'running',
  'finished',
  'canceling',
  'canceled',
  'timing_out',
  'timed_out',
  'skipped',
  'broken'
];

describe('jobTime', () => {
  JOB_STATES.forEach((state) => {
    it(`returns correct time properties for the \`${state}\` state`, () => {
      expect(jobTime({
        state,
        canceledAt: new Date(1475643467000),
        startedAt: new Date(1475643466920),
        finishedAt: new Date(1475643467000),
        timedOutAt: new Date(1475644466920)
      })).toMatchSnapshot();

      expect(jobTime({
        state,
        canceledAt: new Date(1475643467000),
        finishedAt: new Date(1475643467000)
      })).toMatchSnapshot();

      expect(jobTime({
        state,
        timedOutAt: new Date(1475644466920),
        canceledAt: new Date(1475643467000),
        finishedAt: new Date(1475643467000)
      })).toMatchSnapshot();
    });
  });
});