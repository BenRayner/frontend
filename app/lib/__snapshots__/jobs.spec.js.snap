exports[`jobTime returns correct time properties for the \`accepted\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`accepted\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`accepted\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`assigned\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`assigned\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`assigned\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`blocked_failed\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`blocked_failed\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`blocked_failed\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`blocked\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`blocked\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`blocked\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`broken\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`broken\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`broken\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`canceled\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`jobTime returns correct time properties for the \`canceled\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`canceled\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`canceling\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`jobTime returns correct time properties for the \`canceling\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`canceling\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`finished\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T04:57:47.000Z,
}
`;

exports[`jobTime returns correct time properties for the \`finished\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`finished\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`pending\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`pending\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`pending\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`running\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`running\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`running\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`scheduled\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`scheduled\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`scheduled\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`skipped\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`skipped\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`skipped\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`timed_out\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T05:14:26.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`timed_out\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`timed_out\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`timing_out\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
  "to": 2016-10-05T05:14:26.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`timing_out\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`timing_out\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`unblocked_failed\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`unblocked_failed\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`unblocked_failed\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`unblocked\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`unblocked\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`unblocked\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`waiting_failed\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`waiting_failed\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`waiting_failed\` state 3`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`waiting\` state 1`] = `
Object {
  "from": 2016-10-05T04:57:46.920Z,
}
`;

exports[`jobTime returns correct time properties for the \`waiting\` state 2`] = `Object {}`;

exports[`jobTime returns correct time properties for the \`waiting\` state 3`] = `Object {}`;
